File: payola/s2js/runtime/js/s2js/runtime/scala/Array.js
goog.provide('scala.Array');
scala.Array = function(length) {
var self = this;
self.length = length;
self.internalArray = [];};
scala.Array.prototype.s2js_apply = function(i) {
var self = this;
return self.internalArray[i];};
scala.Array.prototype.clone = function() {
var self = this;
return self.internalArray.slice(0);};
scala.Array.prototype.update = function(i, x) {
var self = this;
self.internalArray[i] = x;};
scala.Array.prototype.metaClass_ = new s2js.MetaClass('scala.Array', []);
if (typeof(scala.Array) === 'undefined') { scala.Array = {}; }
scala.Array.fromNative = function(jsArray) {
var self = this;

        var a = new scala.Array(jsArray.length);
        a.internalArray = jsArray;
        return a;
    };
scala.Array.s2js_apply = function() {
var self = this;
var xs = scala.Array.fromNative([].splice.call(arguments, 0, arguments.length - 0));
var array = new scala.Array(xs.length);
var i = 0;
xs.iterator.foreach(function(x) { array.update(i, x);
i = (i + 1);
 });
return array;
};
scala.Array.metaClass_ = new s2js.MetaClass('scala.Array', []);

File: payola/s2js/runtime/src/main/scala/s2js/runtime/scala/Array.scala
package s2js.runtime.scala

import s2js.compiler.NativeJs

object Array {
    @NativeJs("""
        var a = new scala.Array(nativeArray.length);
        a.internalArray = nativeArray;
        return a;
    """)
    def fromNative(nativeArray: AnyRef): Array = null

    @NativeJs("return self.fromNative(xs.internalArray);")
    def apply(xs: Any*): Array = null
}

class Array(val length: Int) {
    @NativeJs("[]")
    var internalArray = null

    @NativeJs("return self.internalArray[i];")
    def apply(i: Int): Any = null

    @NativeJs("return self.internalArray.slice(0);")
    override def clone(): Array = null

    @NativeJs("self.internalArray[i] = x;")
    def update(i: Int, x: Any) {}

    @NativeJs("""self.internalArray.forEach(f);""")
    def foreach(f: Any => Unit) {}
}
